pennylane/math/quantum.py

```diff
@@ -455,3 +456,123 @@ def to_density_matrix(state, indices, check_state=False, c_dtype="complex64"):
     density_matrix = _density_matrix_from_matrix(state, indices, check_state)
 
     return density_matrix
+
+
+def to_vn_entropy(state, indices=None, base=None, check_state=False, c_dtype="complex128"):
+    """Compute the Von Neumann entropy from a state vector, a density matrix given a subsystem.
+
+    Args:
+        state (tensor_like): ``(2**N)`` tensor state vector or ``(2**N, 2**N)`` tensor density matrix.
+        indices (list(int)): List of indices in the considered subsystem.
+        base (float, int): Base for the logarithm.
+        check_state (bool): If True, the function will check the state validity (shape and norm).
+        c_dtype (str): Complex floating point precision type.
+
+    Returns:
+        float: Von Neumann entropy of the considered subsystem.
+
+    **Example**
+
+    >>> x = [1, 0, 0, 1] / np.sqrt(2)
+    >>> to_vn_entropy(x, indices=[0])
+    0.6931472
+
+    >>> to_vn_entropy(x, indices=[0], base=2)
+    1.0
+
+    >>> y = [[1/2, 0, 0, 1/2], [0, 0, 0, 0], [0, 0, 0, 0], [1/2, 0, 0, 1/2]]
+    >>> to_vn_entropy(x, indices=[0])
+    0.6931472
+
+    """
+    state = cast(state, dtype=c_dtype)
+    len_state = state.shape[0]
+
+    if state.shape == (len_state,):
+        density_matrix = to_density_matrix(state, indices, check_state)
+        entropy = compute_vn_entropy(density_matrix, base)
+        return entropy
+
+    density_matrix = to_density_matrix(state, indices, check_state)
+    entropy = compute_vn_entropy(density_matrix, base)
+
+    return entropy
+
+
+def compute_vn_entropy(density_matrix, base=None):
+    """Compute the Von Neumann entropy from a density matrix
+
+    Args:
+        density_matrix (tensor_like): ``(2**N, 2**N)`` tensor density matrix for an integer `N`.
+        base (float, int): Base for the logarithm.
+
+    Returns:
+        float: Von Neumann entropy of the density matrix.
+
+    **Example**
+
+    >>> x = [[1/2, 0], [0, 1/2]]
+    >>> compute_vn_entropy(x, indices=[0])
+    0.6931472
+
+    >>> x = [[1/2, 0], [0, 1/2]]
+    >>> compute_vn_entropy(x, indices=[0], base=2)
+    1.0
+
+    """
+    # Change basis if necessary
+    if base:
+        div_base = np.log(base)
+    else:
+        div_base = 1
+
+    interface = get_interface(density_matrix)
+

```

Comment:
I'm kind of conflicted about having a separate case for each interface. Have you already looked into using functions from `qml.math`? It seems something like the following could work (untested):
```python
evs = qml.math.linalg.eigvalsh(density_matrix)
evs = qml.math.where(evs > 0, evs, 1)
entropy = -qml.math.sum(evs * qml.math.log(evs)) / log_base
```

Reply:
We wanted to use jax and torch function for entropy, therefore I created it like that but I will give it a try. And also autograd is not compatible with `qml.math.linalg.eigvalsh(density_matrix)` unfortunately.

Reply:
If `qml.math.linalg.eigvalsh` doesn't work with autograd, then maybe we can rewrite it in `single_dispatch.py`.

What's the reason for using the JAX and torch functions for entropy? I'd prefer readability from combining all the interfaces over a small performance hit.

Reply:
I will try something and see if the different tests are still passing 👍

Reply:
An interface independent approach sounds nice if possible 👍

Reply:
Done 👍